### PR TITLE
Add checking of patterns in body

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

File: lib/elixir/lib/kernel/special_forms.ex [hunk 1]
```diff
@@ -206,6 +200,12 @@ defmodule Kernel.SpecialForms do
       iex> <<"foo"::utf32>>
       <<0, 0, 0, 102, 0, 0, 0, 111, 0, 0, 0, 111>>
 
+  Otherwise we get an `ArgumentError` when construcing the binary:
+
+      > rest = "oo"
```
Comment (josevalim) [inline lib/elixir/lib/kernel/special_forms.ex]:
```suggestion
      rest = "oo"
```

File: lib/elixir/lib/kernel/special_forms.ex [hunk 2]
```diff
@@ -206,6 +200,12 @@ defmodule Kernel.SpecialForms do
...
+      > <<102, rest>>
```
Comment (josevalim) [inline lib/elixir/lib/kernel/special_forms.ex]:
```suggestion
      <<102, rest>>
```

File: lib/elixir/lib/module/types/expr.ex [hunk 3]
```diff
@@ -1,7 +1,477 @@
...
+    {:ok, :binary, context}
+  end
+
+  # fn -> ... end
+  def of_expr(literal, _stack, context) when is_function(literal) do
+    {:ok, :dynamic, context}
+  end
+
+  # #PID<...>
+  def of_expr(literal, _stack, context) when is_pid(literal) do
```
Comment (josevalim) [inline lib/elixir/lib/module/types/expr.ex]: I think we can return `:pid` here but `reference` and `port` are not valid AST nodes, so you can remove the two clauses below. :)

File: lib/elixir/lib/module/types/expr.ex [hunk 4]
```diff
@@ -1,7 +1,477 @@
...
+  end
+
+  # ()
+  def of_expr({:__block__, _meta, []}, _stack, context) do
+    {:ok, {:atom, nil}, context}
+  end
+
+  # (expr; expr)
+  def of_expr({:__block__, _meta, exprs}, stack, context) do
+    # Clear expression stack here?
```
Comment (josevalim) [inline lib/elixir/lib/module/types/expr.ex]:
```suggestion
    # TODO: Clear expression stack here?
```

?

File: lib/elixir/lib/module/types/expr.ex [hunk 5]
```diff
@@ -1,7 +1,477 @@
...
+    {:ok, :float, context}
+  end
+
+  # "..."
+  def of_expr(literal, _stack, context) when is_binary(literal) do
+    {:ok, :binary, context}
+  end
+
+  # fn -> ... end
+  def of_expr(literal, _stack, context) when is_function(literal) do
```
Comment (josevalim) [inline lib/elixir/lib/module/types/expr.ex]: I have pushed a commit that removes `is_function` from quoted.

**Comment by ericmj** [comment]
Pushed changes from feedback.

**Pull request merged**